rusk/src/lib/ws.rs

```diff
@@ -76,122 +101,316 @@ async fn listening_loop<N: Network, DB: database::DB>(
                 }
                 let (stream, _) = r.unwrap();
 
-                task::spawn(handle_stream(sources.clone(), stream));
+                task::spawn(handle_stream(sources.clone(), stream, shutdown.resubscribe()));
             }
         }
     }
 }
 
-/// A request sent by the websocket client.
-#[derive(Debug)]
-struct Request {
-    headers: serde_json::Map<String, serde_json::Value>,
-    target_type: u8,
-    target: String,
-    topic: String,
-    data: Vec<u8>,
-}
+/// Receives a stream, performs the WebSocket handshake on it, and keeps
+/// receiving messages until either the stream closes, or the server shuts down.
+///
+/// If a message is received that the server cannot parse, then the stream will
+/// be closed.
+async fn handle_stream<VM, S>(
```

Comment:
Shouldn't we limit the number of active streams/connections together with number of active requests per stream to mitigate any DDoS attack?

Reply:
I think you're right. Any suggestions on how to do it?

Reply:
A rate-limiter that is executed as middleware handler for each request may be an option here. This together with max_connections_num config may be a good starting point to secure it. However, this should be addressed in a separate issue.